### PR TITLE
Generalize Cosmos address to alternative pubkey hash, merge evmos

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

Generalize Cosmos address to support alternative pubkey hash, from config; merge native Evmos implementation with Cosmos.  NativeEvmos thus does not need a custom blockchain type, nor Entry, Address implementation.

## How to test

All unit tests, esp. Cosmos-based chains.

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [x] Create pull request as draft initially, unless its complete.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description.

If you're adding a new blockchain

- [ ] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain.

### CODE REVIEW

File: registry.json [hunk 1]
```diff
@@ -2405,6 +2412,7 @@
     "curve": "secp256k1",
     "publicKeyType": "secp256k1Extended",
     "hrp": "evmos",
+    "publicKeyHasher": "keccak256",
```
Comment (hewigovens) [inline registry.json:2415]: This needs to add to all EVM items, not just Evmos

Comment (catenocrypt) [reply registry.json:2415]: Ethereum implementation does not use publicKeyHasher field, now only Cosmos implementation was extended to use this field.  Other than Cosmos, only Bitcoin uses it.

Comment (hewigovens) [reply registry.json:2415]: Yes, but this issue is existing for a while: developer might call `TW::publicKeyHasher(TWCoinTypeEthereum)` and get the default value `sha256ripemd`, so I think it's a good time to fix it

Comment (catenocrypt) [reply registry.json:2443]: Sure, done.

**Comment by hewigovens** [comment]
It seems that `publicKeyHasher` confuses us, being used as extended public key hasher and public key hasher in address, let's use a separate field in registry json? and keep `publicKeyHasher` as it is.

**Comment by catenocrypt** [comment]
Suggestion:
- leave `publicKeyHasher` as it is, used in HDWallet xpub derivation 
- introduce `addressHasher`, used in hashing address, e.g. in Cosmos.  For Eth chains, this can be set to keccak.

File: src/Hash.cpp [hunk 2]
```diff
@@ -19,6 +19,29 @@
...
+        default: case Hash::HashFunc_sha256: return Hash::sha256;
+        case Hash::HashFunc_sha512: return Hash::sha512;
+        case Hash::HashFunc_sha512_256: return Hash::sha512_256;
+        case Hash::HashFunc_keccak256: return Hash::keccak256;
+        case Hash::HashFunc_keccak512: return Hash::keccak512;
+        case Hash::HashFunc_sha3_256: return Hash::sha3_256;
+        case Hash::HashFunc_sha3_512: return Hash::sha3_512;
+        case Hash::HashFunc_ripemd: return Hash::ripemd;
+        case Hash::HashFunc_blake256: return Hash::blake256;
+        //case Hash::HashFunc_blake2b: return static_cast<Hash::HasherSimpleType>(Hash::blake2b);
```
Comment (hewigovens) [inline src/Hash.cpp:34]:
```suggestion
```

File: src/Hash.h [hunk 3]
```diff
@@ -12,10 +12,37 @@
...
+    HashFunc_sha256, // SHA256
+    HashFunc_sha512, // SHA512
+    HashFunc_sha512_256, // SHA512/256
+    HashFunc_keccak256, // Keccak SHA256
+    HashFunc_keccak512, // Keccak SHA512
+    HashFunc_sha3_256, // version 3 SHA256
+    HashFunc_sha3_512, // version 3 SHA512
+    HashFunc_ripemd, // RIPEMD160
+    HashFunc_blake256, // Blake256
+    //HashFunc_blake2b, // Blake2b
```
Comment (hewigovens) [inline src/Hash.h:27]:
```suggestion
```

**Review by hewigovens — APPROVED** [review]
looks good, linux CI is red

**Comment by catenocrypt** [comment]
I have added a few more Hash tests, to restore coverage (Linux CI failed because of coverage).
I also added documentation on fields in `registry.json`:

https://github.com/trustwallet/wallet-core/blob/ar/evmos-cosmos/docs/registry-fields.md

@hewigovens , please re-review.

**Review by hewigovens — APPROVED** [review]

**Pull request merged**